Expand Plugin and REST APIs to trigger user typing event

attilamolnar opened the pull request:
#### Summary
Add endpoint and plugin API to trigger user typing event, see #12609 for more context.

REST endpoint: `/api/v4/users/<userid>/typing`
payload format:
```
{"channel_id":"foo","parent_id":"bar"}
```

This is on top of #14330 as it is needed for the tests.

#### Ticket Link
Fixes #12609 
https://mattermost.atlassian.net/browse/MM-18756

mattermod commented:
Changes detected in `model/client4.go`. Please make sure the https://github.com/mattermost/mmctl is not affected by a breaking change.

mattermod commented:
Thanks @attilamolnar for the pull request!

 Per the [**CONTRIBUTING.md**](https://github.com/mattermost/platform/blob/master/CONTRIBUTING.md#submitting-a-pull-request) file displayed when you created this pull request, we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**

This is a standard procedure for many open source projects. Your form should be processed within 24 hours and reviewers for your pull request will be able to proceed.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

attilamolnar commented:
Would love to get some feedback, especially on
- Access/sanity checks, in which layer to perform them
- Endpoint. Could use `api.BaseRoutes.ChannelForUser` for `/users/<userid>/channels/<channelid>`
- Tests

teresa-novoa commented:
Thank you Attila for your contribution, we’ll start the review process soon! (cc @hanzei)

teresa-novoa commented:
@lieut-data a friendly reminder to help with dev review here, thanks!

lieut-data commented:
Apologies, @attilamolnar -- this draft PR didn't show up in my queue for some reason. I'm going to move this out of a draft PR to help with visibility.

@levb, is this a feature someone on the integrations team might use in an existing/upcoming plugin? Looking to get some sets of eyes on this.

teresa-novoa commented:
Friendly reminder > @levb, is this a feature someone on the integrations team might use in an existing/upcoming plugin? Looking to get some sets of eyes on this. cc @lieut-data Thanks!

lieut-data commented:
@levb, looping you in here to find someone on integrations to review.

lieut-data commented:
@attilamolnar, would you be able to sign the CLA over at https://mattermost.org/mattermost-contributor-agreement/?

levb commented:
Also, @attilamolnar do you mind merging with `master` please and resolving the conflicts?

harshilsharma63 commented:
This is awesome! I've been wanting this feature for a long time.

lieut-data commented:
@attilamolnar / @levb, I've actually gone ahead and merged `master` on your behalf.

I'm 0/5 on the terminology. I do like how it mirrors the `PublishWebSocketEvent` API today. `UserIsTyping` sounds like a hook vs. the API that this is, where we usually have some kind of verb oriented around the caller being the subject.

levb commented:
> I do like how it mirrors the `PublishWebSocketEvent` API today.

Right. That too seems like a valid point for consistency, and it needs to make sense in the client context as well. I definitely don't want to hold up this PR because of minor naming issues, but let me try something "minimalistic" - maybe just `UserTyping`, or even `Typing`, to match the route? Who else can be typing but the User? in MS Teams I think I've seen bots do "typing..." but that just seems hockey.

lieut-data commented:
I actually thought triggering bots typing was the primary motivation for this feature :D

levb commented:
> I actually thought triggering bots typing was the primary motivation for this feature :D

Waking up... sorry for my confusion, all good.

lieut-data commented:
I've submitted an API change to match this that will resolve the build issue once merged.

hanzei commented:
/update-branch

hanzei commented on plugin/api.go:
```diff
@@ -932,6 +932,12 @@ type API interface {
 	// Minimum server version: 5.18
 	PluginHTTP(request *http.Request) *http.Response
+
+	// PublishUserTyping publishes a user is typing WebSocket event.
```
The documentation should make more clear that `parentId` is an optional parameter.

hanzei commented on wsapi/user.go:
```diff
@@ -35,15 +35,9 @@ func (api *API) userTyping(req *model.WebSocketRequest) (map[string]interface{},
 
-	omitUsers := make(map[string]bool, 1)
-	omitUsers[req.Session.UserId] = true
+	appErr := api.App.PublishUserTyping(req.Session.UserId, channelId, parentId)
```
This _might_ be one of the hot areas of the code. Doing two additional DB requests to fetch user and channel could have an high impact. I'm wondering if the readability is really worth the risk of performance issues here.

lieut-data replied:
That's a really good point, @hanzei! Yes, I think we should avoid the extra database read here. In fact, since the existing ws code and the new api4 code already do permission checking, it seems entirely redundant to check that the user and channel are valid at all. That is, I think it's fine to allow a plugin to emit values that "don't make sense" in this context: it's not, strictly speaking, wrong: just meaningless.

So maybe we can remove the two database checks altogether?

attilamolnar replied:
Added a commit https://github.com/mattermost/mattermost-server/pull/14331/commits/20e86f46e4be8a1bf3be982ac134e8bfd5e8112f to remove check. Is this what you had in mind?

hanzei commented on api4/user_test.go:
```diff
@@ -4628,3 +4628,78 @@ func TestGetKnownUsers(t *testing.T) {
+
+	th.Server.Busy.Set(time.Second * 10)
+
+	t.Run("should return service unavailable for non-system admin user when triggering a typing event and server busy", func(t *testing.T) {
+		_, resp := th.Client.PublishUserTyping("invalid", tr)
+		CheckErrorMessage(t, resp, "api.context.server_busy.app_error")
+		CheckServiceUnavailableStatus(t, resp)
+	})
+
+	t.Run("should return service unavailable for system admin user when triggering a typing event and server busy", func(t *testing.T) {
+		_, resp := th.SystemAdminClient.PublishUserTyping(th.SystemAdminUser.Id, tr)
+		CheckErrorMessage(t, resp, "api.context.server_busy.app_error")
+		CheckServiceUnavailableStatus(t, resp)
+	})
```
Great tests :tada:

attilamolnar commented:
Thanks for the comments. I addressed the points you raised in the newly pushed commits.

If all good let me know and I'll go ahead and update the example plugin and docs (if any still todo).

attilamolnar commented:
Have you considered this?

>- Endpoint. Could use `api.BaseRoutes.ChannelForUser` for `/users/<userid>/channels/<channelid>`

lieut-data commented:
/update-branch

prapti commented:
@hanzei Pinging you for some initial test steps as I'm not entirely sure what/how I'm testing, some general steps/directions will do. Thank you in advance!

hanzei commented:
@prapti I did add QA test steps to the JIRA. Let me know if anything is unclear or if you need additional  help.

lieut-data commented:
/update-branch

hanzei commented:
Great work @attilamolnar :+1:

Given that this PR added a new REST API endpoint, this endpoint also has to be documented at https://api.mattermost.com/. Would you be interested in taking care of that?

attilamolnar commented:
@hanzei Thanks. I'm going to update the example plugin next because @lieut-data took care of updating the API docs in mattermost/mattermost-api-reference#526.

hanzei commented:
Oh, I didn't noticed that the docs are already updates. Thanks for reminding me about that @attilamolnar
